handler/handler.go

```diff
@@ -44,7 +76,128 @@ func (h *ReindexRequestedHandler) Handle(ctx context.Context, event *models.Rein
+	datasetURLChan := getAndSendDatasetURLsFromLatestMetadata(ctx, datasetAPICli, editionChan, cfg.ServiceAuthToken)
+	logExtractedDatasetURLs(ctx, datasetURLChan)
+}
+
+func extractDatasets(ctx context.Context, datasetAPIClient clients.DatasetAPIClient, serviceAuthToken string) chan dataset.Dataset {
+	datasetChan := make(chan dataset.Dataset)
+	go func() {
+		defer close(datasetChan)
+		var offset = 0
+		var totalDocs = 0
+		for {
+			list, err := datasetAPIClient.GetDatasets(ctx, "", serviceAuthToken, "", &dataset.QueryParams{
+				Offset: offset,
+				Limit:  DefaultPaginationLimit,
+			})
+			if err != nil {
+				log.Error(ctx, "failed to get dataset clients: %v", err)
+				break
+			}
+
+			listLength := len(list.Items)
+
+			if listLength == 0 {
+				break
+			}
+
+			for i := 0; i < listLength; i++ {
+				datasetChan <- list.Items[i]
+			}
+
+			offset += DefaultPaginationLimit
+			totalDocs += listLength
+		}
+		log.Info(ctx, "total number of dataset docs retrieved", log.Data{"total_documents": totalDocs})
+	}()
+	return datasetChan
+}
+
+func retrieveDatasetEditions(ctx context.Context, datasetAPIClient clients.DatasetAPIClient, datasetChan chan dataset.Dataset, serviceAuthToken string) chan DatasetEditionMetadata {
+	editionMetadataChan := make(chan DatasetEditionMetadata)
+	go func() {
+		defer close(editionMetadataChan)
+		noOfConcurrentExtractions := getNoOfConcurrentExtractions(len(datasetChan), maxConcurrentExtractions)
+		var wg sync.WaitGroup
+		for i := 0; i < noOfConcurrentExtractions; i++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				for dataset := range datasetChan {
+					if dataset.Current == nil {
+						continue
+					}
+					editions, err := datasetAPIClient.GetFullEditionsDetails(ctx, "", serviceAuthToken, dataset.CollectionID, dataset.Current.ID)
+					if err != nil {
+						logData := log.Data{
+							"dataset_id":    dataset.Current.ID,
+							"collection_id": dataset.CollectionID,
+						}
+						log.Error(ctx, "error retrieving editions for dataset", err, logData)
+					}
+					for i := 0; i < len(editions); i++ {
+						if editions[i].ID == "" || editions[i].Current.Links.LatestVersion.ID == "" {
+							continue
+						}
+						editionMetadataChan <- DatasetEditionMetadata{
+							id:        dataset.Current.ID,
+							editionID: editions[i].Current.Edition,
+							version:   editions[i].Current.Links.LatestVersion.ID,
+						}
+					}
+				}
+			}()
+		}
+		wg.Wait()
+	}()
+	return editionMetadataChan
+}
+
+func getAndSendDatasetURLsFromLatestMetadata(ctx context.Context, datasetAPIClient clients.DatasetAPIClient, editionMetadata chan DatasetEditionMetadata, serviceAuthToken string) chan string {
+	datasetURLChan := make(chan string)
+	go func() {
+		defer close(datasetURLChan)
+		var wg sync.WaitGroup
+		noOfConcurrentExtractions := getNoOfConcurrentExtractions(len(editionMetadata), maxConcurrentExtractions)
+		for i := 0; i < noOfConcurrentExtractions; i++ {
```

Comment:
How big can 'noOfConcurrentExtractions' become ?
You don't have unlimited resources ?
This kind of thing should have some sort of limit and the code would process chunks at a time.
(just a thought to bear in mind for the future)

Reply:
**Number of go-routines in the `Handle` function for `reindex-requested`**
2 - Handle func
1 - extractDatasets func
1 + maxConcurrentExtractions (20) - retrieveDatasetEditions func
1 + maxConcurrentExtractions (20) - getAndSendDatasetURLsFromLatestMetadata func

So currently, there are at most 45 goroutines for this particular functionality.
So the number of go-routines for `Handle` can be equated as `5 + (2 * maxConcurrentExtractions)`

Whilst looking online (https://stackoverflow.com/questions/8509152/max-number-of-goroutines and https://calmops.com/golang/golang-limit-total-number-of-goroutines/), a good rule of thumb is to have less 10,000 goroutines (please note that we can have more depending on memory and the functionality of the app) but for now, this is good because we are not even utilizing close to 10000 go-routines in the overall app. 

Doing the inverse of the equated func, it is best for the `maxConcurrentExtractions` to be at max 4000 ((10000 - 5) / 2 = 4997.5 and then leave a good buffer to cater for other go-routines in the app)

Reply:
Hmmm ... thats all theory.
You have to take in to account the size of machine being used to run your app.
Our apps are run out of a container managed by nomad and we run many apps on one quite small box, so each app typically only needs a few 100Mb RAM and a few hundred MHz CPU (though some need more).
You do not want an app to go exploding its memory requirements, because we have our machines set up to allow bursting through CPU limits, which then squeezes out other apps on the same box.
You muast always consider your available resources - they are not unlimited.
Look into the runtime environment of your apps and figure out the constraints.
Also with 1,000 of go routines, they need garbage collection management which is nit for free !